frontend/app/assets/javascripts/spree/frontend/cart.js

```diff
@@ -18,100 +21,138 @@ Spree.ready(function ($) {
     return triggerObject
   }
 
-  if (formUpdateCart.length) {
-    var clearInvalidCouponField = function() {
-      var couponCodeField = $('#order_coupon_code');
-      var couponStatus = $('#coupon_status');
-      if (!!couponCodeField.val() && couponStatus.hasClass('alert-error')) {
-        couponCodeField.val('')
+  function getLineItemId(element) {
+    return $(element).attr('data-id').replace('line_item_', '')
+  }
+
+  function handleCartApiError(error, target) {
+    if (target) target.attr('disabled', false)
+    Spree.hideProgressBar()
+    alert(error)
+  }
+
+  function handleCartApiSuccess() {
+    window.location.reload()
+  }
+
+  function handleSetQuantity(lineItemId, quantity, input, target) {
+    Spree.showProgressBar()
+    target.attr('disabled', 'true')
+    input.val(quantity)
+
+    SpreeAPI.Storefront.setLineItemQuantity(
+      lineItemId,
+      quantity,
+      function(response) { handleCartApiSuccess() },
+      function(error) {
+        handleCartApiError(error, target)
+        input.val(quantity - 1) // revert to previous number
       }
+    )
+  }
+
+  if (formUpdateCart.length) {
+    var COUPON_CODE_ELEMENTS = {
+      appliedCouponCodeField: formUpdateCart.find('#order_applied_coupon_code'),
+      couponCodeField: formUpdateCart.find('#order_coupon_code'),
+      couponStatus: formUpdateCart.find('#coupon_status'),
+      couponButton: formUpdateCart.find('#shopping-cart-coupon-code-button'),
+      removeCouponButton: formUpdateCart.find('#shopping-cart-remove-coupon-code-button')
     }
 
+    // handle remove line item from cart
     formUpdateCart.find('a.delete').show().one('click', function (event) {
-      var itemId = $(this).attr('data-id')
-      var link = $(event.currentTarget);
-      var quantityInputs = $("form#update-cart input.shopping-cart-item-quantity-input[data-id='" + itemId + "']")
+      event.preventDefault()
+
+      var lineItemId = getLineItemId(this)
+      var button = $(event.target)
+      // FIXME: this selector madness need to go away...
+      var quantityInputs = $("form#update-cart input.shopping-cart-item-quantity-input[data-id='line_item" + lineItemId + "']")
```

Comment:
I think we can simplify it as following:
```suggestion
      var quantityInputs = document.querySelector(`input[data-id='line_item_${lineItemId}']`)
```